src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs

```diff
@@ -258,6 +263,639 @@ public static async Task LargeJsonFile(int bufferSize)
+            int length = 10 * depthFactor;
+            List<Post>[] posts = new List<Post>[length];
+            posts[0] = PopulateLargeObject(1);
+            for (int i = 1; i < length; i++ )
+            {
+                posts[i] = PopulateLargeObject(1);
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = ignoreNull,
+                WriteIndented = writeIndented
+            };
+            string json = JsonSerializer.Serialize(posts[0], options);
+
+            // Sync case.
+            {
+                List<Post> deserializedList = JsonSerializer.Deserialize<List<Post>>(json, options);
+
+                string jsonSerialized = JsonSerializer.Serialize(deserializedList, options);
+                Assert.Equal(json, jsonSerialized);
+            }
+
+            // Async case.
+            using (var memoryStream = new MemoryStream())
+            {
+                await JsonSerializer.SerializeAsync(memoryStream, posts[0], options);
+                string jsonSerialized = Encoding.UTF8.GetString(memoryStream.ToArray());
+                Assert.Equal(json, jsonSerialized);
+
+                memoryStream.Position = 0;
+                List<Post> deserializedList = await JsonSerializer.DeserializeAsync<List<Post>>(memoryStream, options);
+            }
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(16)]
+        public static async Task DeepNestedJsonFileCircularDependencyTest(int depthFactor)
+        {
+            int length = 10 * depthFactor;
+            List<Post>[] posts = new List<Post>[length];
+            posts[0] = PopulateLargeObject(1000);
+            for (int i = 1; i < length; i++)
+            {
+                posts[i] = PopulateLargeObject(1);
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+            posts[length - 1][0].Value.PrimaryTopic.RelatedTopics = posts[0];
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = true
+            };
+
+            Assert.Throws<JsonException> (() => JsonSerializer.Serialize(posts[0], options));
+
+            using (var memoryStream = new MemoryStream())
+            {
+                await Assert.ThrowsAsync<JsonException>(async () => await JsonSerializer.SerializeAsync(memoryStream, posts[0], options));
+            }
+        }
+
+
+        [Theory]
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int )(bufferSize * 0.9 - 2);
```

Comment:
`bufferSize * 0.9` is threshold size from codebase - 2 for `["` characters, after these chars we add that amount of char `a` which produces buffer written equal to threshold size, after this long string adding a big object to produce scenario noted in: https://github.com/dotnet/corefx/issues/39500 :

> It would also be good to add tests where depth changes at those size threshold (for example, writing a start or end array/object at the 16k bytes)

Reply:
Please add that as a comment within the test.